pkg/subsystem/syz-query-subsystems/generator.go

```diff
@@ -0,0 +1,184 @@
+// Copyright 2023 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"go/format"
+	"log"
+	"regexp"
+	"sort"
+	"strings"
+	"text/template"
+
+	"github.com/google/syzkaller/pkg/serializer"
+	"github.com/google/syzkaller/pkg/subsystem/entity"
+)
+
+func generateSubsystemsFile(name string, subsystems []*entity.Subsystem) ([]byte, error) {
+	// Drop subsystems without a name.
+	list := []*entity.Subsystem{}
+	for _, entry := range subsystems {
+		if entry.Name == "" {
+			log.Printf("skipping a subsystem with no name (lists: %#v, rules: %#v)", entry.Lists, entry.PathRules)
+			continue
+		}
+		// Clean parents.
+		newParents := []*entity.Subsystem{}
+		for _, p := range entry.Parents {
+			if p.Name != "" {
+				newParents = append(newParents, p)
+			}
+		}
+		entry.Parents = newParents
+		list = append(list, entry)
+	}
+
+	// Sort subsystems by name to keep output consistent.
+	sort.Slice(list, func(i, j int) bool { return list[i].Name < list[j].Name })
+
+	// Sort path rules to keep output consistent.
+	for _, entity := range list {
+		sort.Slice(entity.PathRules, func(i, j int) bool {
+			a, b := entity.PathRules[i], entity.PathRules[j]
+			if a.IncludeRegexp != b.IncludeRegexp {
+				return a.IncludeRegexp < b.IncludeRegexp
+			}
+			return a.ExcludeRegexp < b.ExcludeRegexp
+		})
+	}
+
+	// Set names first -- we'll need them for filling in the Parents array.
+	objToName := map[*entity.Subsystem]string{}
+	for _, entry := range list {
+		varName := makeVarRegexp.ReplaceAllString(strings.ToLower(entry.Name), "")
+		if varName == "" {
+			return nil, fmt.Errorf("failed to get a var name for %#v", entry.Name)
+		}
+		objToName[entry] = varName
+	}
+
+	// Prepare the template data.
+	vars := &templateVars{
+		Name:        name,
+		PackageName: strings.ReplaceAll(name, "_", ""),
+		Hierarchy:   hierarchyList(list),
+	}
+	for _, entry := range list {
+		varName := objToName[entry]
+		// The serializer does not understand parent references and just prints all the
+		// nested structures.
+		// Therefore we call it separately for the fields it can understand.
+		parents := []string{}
+		for _, p := range entry.Parents {
+			parents = append(parents, objToName[p])
+		}
+		sort.Strings(parents)
+		subsystem := &templateSubsystem{
```

Comment:
It may be simpler if we make the list of subsystems `map[string]Subsystem` and use strings to reference parents. Have you considered such option?

Reply:
I considered it before writing all the code, but did not find that one is sufficiently better than another. Most of the code doesn't really care about names and, except for a few cases, pointers are just fine.

If we use `map[string]Subsystem`, we have to pass the map everywhere and in every place do corner case checks like whether the specified parent name is really present in the map etc. We don't have to do it when using pointers (but yes, the code generation gets a bit more difficult).

There was also a small argument against using the `map[string]Subsystem` approach -- in case of conflicts, not all subsystems could get a name. It seemed a better approach to let the caller of `pkg/subsystem` decide what to do (e.g. skip empty ones and report the error somehow -- it's easier to do it from dashboard or syz-ci than from pkg/subsystem). At the time when the decision was made I was expecting the subsystem extraction to happen automatically and this problem was very relevant.

Now that we'll be doing such updates by a manual invocation of the `syz-query-subsystems` tool, it's no longer that important.

So neither of the approaches is in my view necessarily better, and the question is whether it's now worth the effort to rewrite the already existing implementation.